src/inc/til/io.h

```diff
@@ -0,0 +1,273 @@
+                if (elevatedOnly)
+                {
+                    const auto hadExpectedPermissions{ details::isOwnedByAdministrators(file.get()) };
+                    if (!hadExpectedPermissions)
+                    {
+                        // Close the handle
+                        file.reset();
+
+                        // delete the file. It's been compromised.
+                        LOG_LAST_ERROR_IF(!DeleteFile(path.c_str()));
+
+                        // Exit early, because obviously there's nothing to read from the deleted file.
+                        return "";
+                    }
+                }
+
+                const auto fileSize = GetFileSize(file.get(), nullptr);
+                THROW_LAST_ERROR_IF(fileSize == INVALID_FILE_SIZE);
+
+                // By making our buffer just slightly larger we can detect if
+                // the file size changed and we've failed to read the full file.
+                std::string buffer(static_cast<size_t>(fileSize) + 1, '\0');
+                DWORD bytesRead = 0;
+                THROW_IF_WIN32_BOOL_FALSE(ReadFile(file.get(), buffer.data(), gsl::narrow<DWORD>(buffer.size()), &bytesRead, nullptr));
+
+                // This implementation isn't atomic as we'd need to use an exclusive file lock.
+                // But this would be annoying for users as it forces them to close the file in their editor.
+                // The next best alternative is to at least try to detect file changes and retry the read.
+                if (bytesRead != fileSize)
+                {
+                    // This continue is unlikely to be hit (see the prior for loop comment).
+                    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                    continue;
+                }
+
+                // As mentioned before our buffer was allocated oversized.
+                buffer.resize(bytesRead);
+
+                if (til::starts_with(buffer, details::Utf8Bom))
+                {
+                    // Yeah this memmove()s the entire content.
+                    // But I don't really want to deal with UTF8 BOMs any more than necessary,
+                    // as basically not a single editor writes a BOM for UTF8.
+                    buffer.erase(0, details::Utf8Bom.size());
+                }
+
+                if (lastWriteTime)
+                {
+                    THROW_IF_WIN32_BOOL_FALSE(GetFileTime(file.get(), nullptr, nullptr, lastWriteTime));
+                }
+
+                return buffer;
+            }
+
+            THROW_WIN32_MSG(ERROR_READ_FAULT, "file size changed while reading");
+        }
+
+        // Same as read_file_as_utf8_string, but returns an empty optional, if the file couldn't be opened.
+        _TIL_INLINEPREFIX std::optional<std::string> read_file_as_utf8_string_if_exists(const std::filesystem::path& path, const bool elevatedOnly, FILETIME* lastWriteTime)

```

Comment:
Checking all calls to `read_file_as_utf8_string_if_exists` all of them follow up with a `.value_or(std::string{})`. I think we should just return a `std::string`.

Furthermore I found that there's only a single caller to `read_file_as_utf8_string`. That caller doesn't care if the file could not be found. As such it can be replaced with a call to `read_file_as_utf8_string_if_exists` and just check if the returned string is empty.

This drops the need for having 2 variants of this function.

Reply:
neat! I should probably do this

Reply:
Do you still want to do this?

Reply:
I DO I AM JUST LAZY

Reply:
![image](https://github.com/user-attachments/assets/813bdc69-8188-4c4f-beb1-a1c573ebf340)


we may be too lazy to do this after all

Reply:
IT ME
I'M THE LAZY ONE